### PR TITLE
Initial working version

### PR DESCRIPTION
**Pull request by dcheckoway**
First working version of the plugin.  Much of this is boilerplate stuff, created with `npx @grafana/create-plugin@latest`.  The moving parts that are Cribl-specific (where you should focus the code review) are all in `src`:

- `src/README.md` (info about the plugin that shows up in Grafana, provided by Emil)
- `src/plugin.json` (definitions, metadata, proxied API routes, etc.)
- `src/module.ts` ("entry point" where the plugin is defined & exported)
- `src/types.ts` (shared custom types)
- `src/datasource.ts` (this is where it's at)
- `src/components/ConfigEditor.tsx` (UI for configuring the plugin)
- `src/components/QueryEditor.tsx` (UI for querying -- this will definitely evolve over time)

A note to the uninitiated about how things work in the Grafana plugin ecosystem:

- The file naming convention is Grafana's.  The mix of all-lower and camel-case file names, that's auto-generated by their plugin creator.  No reason to deviate, we'll keep things as compliant as possible to streamline the approval/publishing process.
- IMPORTANT to keep in mind: This all runs _in the browser_, not on the Grafana backend.
- You'll see API calls being made with `getBackendSrv().fetch()`, which ordinarily makes HTTP requests directly from the browser -- but in our case we've defined "routes" in `plugin.json`.  You'll see the `proxiedUrl()` function in `datasource.ts` taking advantage of this, proxying various API calls through the Grafana backend.

### CODE REVIEW

**Comment by dcheckoway** [comment]
@ollidraese `LICENSE` in the root dir is Apache 2.0.  That was spat out by the plugin generator by default.  Maybe you can take a peek when you have a sec, lemme know if you feel we're covered?  TIA!

File: src/types.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+   */
+  maxResults: number;
+}
+
+/**
+ * Options configured for each CriblDataSource instance
+ */
+export interface CriblDataSourceOptions extends DataSourceJsonData {
+  /**
+   * Base URL to the Cribl organization/tenant site (i.e. https://main-elated-nash-dkf9udv.cribl-staging.cloud)
```
Comment (bardhi) [inline src/types.ts:23]: nit: might be nice to use a non-staging URL as an example. I'd suggesting removing the `-staging` part (no need for this to be a REAL tenant url)

File: .config/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
+ARG grafana_version=latest
+ARG grafana_image=grafana-enterprise
```
Comment (ollidraese) [inline .config/Dockerfile:2]: shouldn't we default to the community edition here?

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+dist/
+artifacts/
+work/
+ci/
+e2e-results/
+**/cypress/videos
+**/cypress/report.json
+
+# Editor
+.idea
```
Comment (ollidraese) [inline .gitignore:33]: how about ignores of other editors (i.e. ,.vscode)?

File: .nvmrc [hunk 4]
```diff
@@ -0,0 +1 @@
+16
```
Comment (ollidraese) [inline .nvmrc:1]: this is a new project and 16 is getting out of support in September. Can we bump this (i.e. to 20)?

File: CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# Changelog
+
+## 1.0.0
+
+Initial release.
```
Comment (ollidraese) [inline CHANGELOG.md:5]: a little more context (what is part of this initial release) would be gr8

File: LICENSE [hunk 6]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (ollidraese) [inline LICENSE:1]: q: is this in sync w/ other OS projects that we have or do we have a mix of i.e. Apache vs. MIT?

File: README.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
+# Cribl Search Grafana Plugin
+
+This is a Data Source Plugin for Grafana that pulls from Cribl Search.
```
Comment (ollidraese) [inline README.md:3]: What is Cribl Search? How does it get triggered? Link to product? Samples?

File: src/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
+# Grafana Cribl Plugin
+
+Welcome to the Grafana Cribl plugin README. This plugin empowers you to create Grafana data sources pulling data directly from Cribl Search, with the resulting data visualizations offering insights like never before. Let's get into how to set this up.
```
Comment (ollidraese) [inline src/README.md:3]: great. I would have expected most of this in the top-level README. Is this README at least linked from there?

File: src/components/QueryEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+  const onMaxResultsChange = (event: ChangeEvent<HTMLInputElement>) => {
+    onChange({ ...query, maxResults: +event.target.value });
+    debouncedOnRunQuery();
+  };
+
+  const { savedQueryId, maxResults } = query;
+
+  return (
+    <div className="gf-form">
```
Comment (ollidraese) [inline src/components/QueryEditor.tsx:26]: Any chance that we could add test/verify here?

- is the saved-queryi-d valid?
- do I have access (ACL) to the saved query?

File: src/datasource.ts [hunk 10]
```diff
@@ -0,0 +1,300 @@
+import {
```
Comment (ollidraese) [inline src/datasource.ts:1]: Doesn't Apache expect you to put a license reference in the headers of sources?
https://www.apache.org/legal/src-headers.html

File: src/datasource.ts [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+  private async processQuery(criblQuery: CriblQuery): Promise<DataFrame> {
+    let fields: Record<string, Field> = {};
+    let eventCount = 0;
+    let totalEventCount: number | undefined = undefined;
+    let maxEventCount = criblQuery.maxResults ?? Number.MAX_VALUE;
+
+    // Load the search results, paging through until we've hit maxResults or read them all, whatever comes first
+    do {
+      const response = await this.getAuthorization() // refresh auth token as needed
+        .then((authorization) => lastValueFrom(getBackendSrv().fetch({
```
Comment (ollidraese) [inline src/datasource.ts:52]: nit: not sure how much I like this combination of await and promise-chaining via .then (btw: no .catch). Could we stick to await only and make this code look sequential?

File: src/datasource.ts [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+  /**
+   * Run a single query to produce a single DataFrame
+   * @param criblQuery the query to run
+   * @returns the DataFrame with the results
+   */
+  private async processQuery(criblQuery: CriblQuery): Promise<DataFrame> {
+    let fields: Record<string, Field> = {};
+    let eventCount = 0;
+    let totalEventCount: number | undefined = undefined;
+    let maxEventCount = criblQuery.maxResults ?? Number.MAX_VALUE;
```
Comment (ollidraese) [inline src/datasource.ts:47]: do we need a sanity check on criblQuery.maxResults (i.e.> 1)?

File: src/datasource.ts [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+          headers: { ...authorization },
+          responseType: 'text', // tell Grafana not to bother trying to JSON parse it, it's NDJSON
+          params: {
+            queryId: criblQuery.savedQueryId,
+            offset: eventCount,
+            limit: Math.min(maxEventCount, QUERY_PAGE_SIZE),
+          },
+        })));
+      if (response.status !== 200) {
+        throw new Error(`Unexpected response status (${response.status})`);
```
Comment (ollidraese) [inline src/datasource.ts:64]: q: does the endpoint also return 200 on empty results?

File: src/datasource.ts [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+      }
+
+      // Parse the NDJSON lines
+      const data = response.data as string;
+      let lines = data.split('\n');
+      for (let lineIdx = 0; lineIdx < lines.length && eventCount < maxEventCount; ++lineIdx) {
+        if (lines[lineIdx].length === 0) {
+          continue;
```
Comment (ollidraese) [inline src/datasource.ts:72]: q: where do the empty lines come from? trailing NL? comment would be helpful

File: src/datasource.ts [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+  expiresAt: number,
+}
+
+/**
+ * Parse a JWT and get its expiration time
+ * @param token the encoded JWT
+ * @returns the expiration time in epoch seconds
+ */
+function parseJwtExp(token: string): number {
+  const b64 = token.split('.')[1].replace(/-/g, '+').replace(/_/g, '/');
```
Comment (ollidraese) [inline src/datasource.ts:295]: q: could we validate the JWT signature here? Is the server really the one we expected to talk to?

**Review by ollidraese — APPROVED** [review]
Would be great if we could (in future version) factor out the client functionality (AuthN, Results fetch, ...) into an independent SDK and reference it from here.

File: .config/Dockerfile (hunk 2 above)
Comment (dcheckoway) [reply .config/Dockerfile:2]: https://grafana.com/docs/grafana/latest/setup-grafana/installation/docker/#run-grafana-docker-image

> Note: The recommended and default edition of Grafana is Grafana Enterprise. It is free and includes all the features of the OSS edition. Additionally, you have the option to upgrade to the [full Enterprise feature set](https://grafana.com/products/enterprise/?utm_source=grafana-install-page), which includes support for [Enterprise plugins](https://grafana.com/grafana/plugins/?enterprise=1&utcm_source=grafana-install-page).

File: .nvmrc (hunk 4 above)
Comment (dcheckoway) [reply .nvmrc:1]: Looks like we can try: https://community.grafana.com/t/which-node-version-to-use/34347/3

But @ollidraese I tried it, and found that it's going to be a slippery slope with needing to upgrade dependencies.  This was the first one on the slope:

```
11:06:38 in cribl-search-grafana-plugin on initial-version [!] on 🐳 v24.0.2 ✘1 ➜ yarn install
yarn install v1.22.19
warning ../../package.json: No license field
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
[3/5] 🚚  Fetching packages...
error eslint-plugin-jsdoc@39.6.2: The engine "node" is incompatible with this module. Expected version "^14 || ^16 || ^17 || ^18 || ^19". Got "20.5.0"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

I think we should get this plugin out the door with 16, and maybe followup with a 20 upgrade shortly down the road.

File: src/README.md (hunk 8 above)
Comment (dcheckoway) [reply src/README.md:3]: This shows up in the Grafana UI when you click on the Cribl plugin.

File: LICENSE (hunk 6 above)
Comment (dcheckoway) [reply LICENSE:1]: I spot-checked a bunch of criblio GitHub repos, and I see a mix of Apache and MIT.

File: src/datasource.ts (hunk 11 above)
Comment (dcheckoway) [reply src/datasource.ts]: Yep, removed the chaining.  FWIW, the `await` here should have precluded the need for an explicit `catch` -- any error thrown at any point in the chain would end up percolating.  But I'm with you, don't want to use a style that isn't preferred by the team, so I changed it.

File: src/datasource.ts (hunk 12 above)
Comment (dcheckoway) [reply src/datasource.ts:47]: I tested it with `0` and it "works as expected" -- arguably useless.  I added `Math.max(1, ...)` to the UI (in `QueryEditor.tsx`).

File: src/datasource.ts (hunk 13 above)
Comment (dcheckoway) [reply src/datasource.ts:64]: Good question...yep!  You still get our "header line" in the NDJSON response, no matter how many results.  Tested, works.

File: src/datasource.ts (hunk 14 above)
Comment (dcheckoway) [reply src/datasource.ts:72]: Added a comment

File: src/components/QueryEditor.tsx (hunk 9 above)
Comment (dcheckoway) [reply src/components/QueryEditor.tsx:26]: Yes and no.  Yes in that we can at least ensure `savedQueryId` isn't _obviously invalid_ (i.e. matches the expected regex), which I'll add.  But no in that the `/search/query` API should take care of enforcing ACL without requiring a separate check for that.

File: src/datasource.ts (hunk 10 above)
Comment (dcheckoway) [reply src/datasource.ts:1]: Nah, that only applies to Apache projects, stuff that's published by ASF.

File: README.md (hunk 7 above)
Comment (dcheckoway) [reply README.md:3]: Added product link.  There are 2 screenshots included in `src/img`, but it (the plugin validator) wouldn't allow me to use relative links to those images.

I'm going to ask the Community about this, about where those screenshots end up if they're not referenced in README.

File: src/README.md (hunk 8 above)
Comment (dcheckoway) [reply src/README.md]: Added a link & more info to the top-level README.

File: README.md (hunk 7 above)
Comment (dcheckoway) [reply README.md:3]: https://community.grafana.com/t/question-about-screenshots/100172

File: .config/Dockerfile (hunk 2 above)
Comment (dcheckoway) [reply .config/Dockerfile:2]: Actually you're right.  `grafana-oss` is what I've been using locally.  I'll update this.

File: .nvmrc (hunk 4 above)
Comment (dcheckoway) [reply .nvmrc]: Upgraded to Node v18

File: README.md (hunk 7 above)
Comment (dcheckoway) [reply README.md:3]: The screenshots end up showing up in the catalog entry.  Can't reference them from README on the plugin page, but they _will_ show up in the catalog at least.

File: README.md [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+## Getting started
+
+
+### Frontend
+
+1. Install dependencies
+
+   ```bash
+   npm run install
```
Comment (bardhi) [inline README.md:26]: Based on the comment above, this is autogenerated... however weirdly this command doesn't work for me, fwiw. I need to do `npm install` instead.

Comment (dcheckoway) [reply README.md:26]: Good catch, I'll update this

**Review by bardhi — APPROVED** [review]

**Pull request merged**